DMD.C

```diff
@@ -0,0 +1,489 @@
+
+#include "Matrix.h"
+#include "Vector.h"
+#include "scalapack_wrapper.h"
+#include "mpi.h"
+
+/* Use C++11 built-in shared pointers if available; else fallback to Boost. */
+#if __cplusplus >= 201103L
+#include <memory>
+#else
+#include <boost/shared_ptr.hpp>
+#endif
+
+#include "scalapack_wrapper.h"
+
+/* Use automatically detected Fortran name-mangling scheme */
+#define dgeev CAROM_FC_GLOBAL(dgeev, DGEEV)
+#define zgetrf CAROM_FC_GLOBAL(zgetrf, ZGETRF)
+#define zgetri CAROM_FC_GLOBAL(zgetri, ZGETRI)
+
+extern "C" {
+    // Compute eigenvalue and eigenvectors of real non-symmetric matrix.
+    void dgeev(char*, char*, int*, double*, int*, double*, double*, double*, int*, double*, int*, double*, int*, int*);
+
+    // LU decomposition of a general matrix.
+    void zgetrf(int*, int*, double*, int*, int*, int*);
+
+    // Generate inverse of a matrix given its LU decomposition.
+    void zgetri(int*, double*, int*, int*, double*, int*, int*);
+}
+
+using namespace std;
+
+namespace CAROM {
+
+DMD::DMD(int dim)
+{
+    // Get the rank of this process, and the number of processors.
+    int mpi_init;
+    MPI_Initialized(&mpi_init);
+    if (mpi_init == 0) {
+        MPI_Init(nullptr, nullptr);
+    }
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &d_rank);
+    MPI_Comm_size(MPI_COMM_WORLD, &d_num_procs);
+    d_dim = dim;
+}
+
+bool DMD::takeSample(double* u_in)
+{
+    CAROM_VERIFY(u_in != 0);
+
+    Vector sample(u_in, d_dim, true);
+    d_snapshots.push_back(sample);
+    return true;
+}
+
+void DMD::train(double energy_fraction)
+{
+    const Matrix* f_snapshots = getSnapshotMatrix();
+    CAROM_VERIFY(f_snapshots->numColumns() > 1);
+    CAROM_VERIFY(energy_fraction > 0 && energy_fraction <= 1);
+    d_energy_fraction = energy_fraction;
+    d_k = f_snapshots->numColumns() - 1;
+    constructDMD(f_snapshots, d_rank, d_num_procs);
+
+    delete f_snapshots;
+}
+
+void DMD::train(int k)
+{
+    const Matrix* f_snapshots = getSnapshotMatrix();
+    CAROM_VERIFY(f_snapshots->numColumns() > 1);
+    CAROM_VERIFY(k > 0 && k <= f_snapshots->numColumns() - 1);
+    d_energy_fraction = -1;
+    d_k = k;
+    constructDMD(f_snapshots, d_rank, d_num_procs);
+
+    delete f_snapshots;
+}
+
+void
+DMD::constructDMD(const Matrix* f_snapshots,
+                  int d_rank,
+                  int d_num_procs)
+{
+    Matrix* f_snapshots_minus = new Matrix(f_snapshots->numRows(),
+                                           f_snapshots->numColumns() - 1, f_snapshots->distributed());
+    Matrix* f_snapshots_plus = new Matrix(f_snapshots->numRows(),
+                                          f_snapshots->numColumns() - 1, f_snapshots->distributed());
+
+    // Break up snapshots into snapshots_minus and snapshots_plus
+    // snapshots_minus = all columns of snapshots except last
+    // snapshots_plus = all columns of snapshots except first
+    for (int i = 0; i < f_snapshots->numRows(); i++)
+    {
+        f_snapshots_minus->item(i, 0) = f_snapshots->item(i, 0);
```

Comment:
This loop could be simplified (just use +1 in "plus" matrix, with `j` starting at 0).